enclave-runtime/src/teeracle/mod.rs

```diff
@@ -42,6 +45,102 @@ use sgx_types::sgx_status_t;
 use sp_runtime::OpaqueExtrinsic;
 use std::{string::String, vec::Vec};
 
+fn update_weather_data_internal(weather_info: WeatherInfo) -> Result<Vec<OpaqueExtrinsic>> {
+	let extrinsics_factory = get_extrinsic_factory_from_solo_or_parachain()?;
+	let ocall_api = GLOBAL_OCALL_API_COMPONENT.get()?;
+
+	let mut extrinsic_calls: Vec<OpaqueCall> = Vec::new();
+
+	let open_meteo_weather_oracle = create_open_meteo_weather_oracle(ocall_api);
+
+	match get_longitude(weather_info, open_meteo_weather_oracle) {
+		Ok(opaque_call) => extrinsic_calls.push(opaque_call),
+		Err(e) => {
+			error!("[-] Failed to get the newest longitude from OpenMeteo. {:?}", e);
+		},
+	};
+	let extrinsics = extrinsics_factory.create_extrinsics(extrinsic_calls.as_slice(), None)?;
+	Ok(extrinsics)
+}
+
+fn get_longitude<OracleSourceType, MetricsExporter>(
+	weather_info: WeatherInfo,
+	oracle: WeatherOracle<OracleSourceType, MetricsExporter>,
+) -> Result<OpaqueCall>
+where
+	OracleSourceType: OracleSource<
+		WeatherInfo,
+		OracleRequestResult = std::result::Result<f32, ita_oracle::error::Error>,
+	>,
+	MetricsExporter: ExportMetrics<WeatherInfo>,
+{
+	let longitude =
+		oracle.get_longitude(weather_info.clone()).map_err(|e| Error::Other(e.into()))?;
+
+	let base_url = oracle.get_base_url().map_err(|e| Error::Other(e.into()))?;
+	let source_base_url = base_url.as_str();
+
+	println!("Update the longitude:  {}, for source {}", longitude, source_base_url);
+
+	let node_metadata_repository = get_node_metadata_repository_from_solo_or_parachain()?;
+
+	let call_ids = node_metadata_repository
+		.get_from_metadata(|m| m.update_oracle_call_indexes())
+		.map_err(Error::NodeMetadataProvider)?
+		.map_err(|e| Error::Other(format!("{:?}", e).into()))?;
+
+	let call = OpaqueCall::from_tuple(&(
+		call_ids,
+		weather_info.weather_query.key().as_bytes().to_vec(),
+		source_base_url.as_bytes().to_vec(),
+		longitude.encode(),
+	));
+
+	Ok(call)
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn update_weather_data_xt(
+	weather_info_longitude: *const u8,
+	weather_info_longitude_size: u32,
+	weather_info_latitude: *const u8,
+	weather_info_latitude_size: u32,
+	unchecked_extrinsic: *mut u8,
+	unchecked_extrinsic_size: u32,
+) -> sgx_status_t {
+	let mut weather_info_longitude_slice =
+		slice::from_raw_parts(weather_info_longitude, weather_info_longitude_size as usize);
+	let longitude =
+		String::decode(&mut weather_info_longitude_slice).expect("Can unwrap into WeatherInfo");
+
+	let mut weather_info_latitude_slice =
+		slice::from_raw_parts(weather_info_latitude, weather_info_latitude_size as usize);
+	let latitude =
+		String::decode(&mut weather_info_latitude_slice).expect("Can unwrap into WeatherInfo");
```

Comment:
I am uncertain if I agree with the `expect` as the gate is coming from an FFI. I'd prefer to return an error here. In general, we must never panic in an enclave as it might expose confidential data.

Reply:
In the other functions we were just calling `.unwrap()` with no message.

Reply:
So was trying to improve upon a simple `unwrap()`